File: leetcode/maximum_subarray/maximum_subarray.java
package maximum_subarray;

class Solution {
    public int maxSubArray(int[] nums) {
        int max = nums[0], maxCurr = nums[0];
        for(int i = 1; i < nums.length; i++) {
            maxCurr = Math.max(maxCurr+nums[i], nums[i]);
            max = Math.max(maxCurr, max);
        }
        return max;
    }
}

File: leetcode/maximum_subarray/maximum_subarray.cpp
/*
 * Time Complexity: O(N)
 *   - N is the length of the array.
 *
 * Space Complexity: O(1)
 *   - Only a constant amount of extra space is used.
 */

#include<vector>
#include<algorithm>

using std::vector;
using std::max;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSub = nums[0], maxCurr = nums[0];
        for(int i = 1; i < nums.size(); i++) {
            maxCurr = max(maxCurr+nums[i], nums[i]);
            maxSub = max(maxCurr, maxSub);
        }
        return maxSub;
    }
};